content/blog/semgrep-post-message/index.md

```diff
@@ -0,0 +1,175 @@
+
+> Since Semgrep is semantic, patterns matching and old `function` declaration will also match code using `arrow functions`
+
+ Now, which patterns do we consider safe for origin checking?
+
+## Safe patterns
+
+If the function specified above contains any of the following patterns we **should not** flag them:
+
+```js
+if (evt.origin == "http://example.com") { ... }
+
+if (evt.origin === "http://example.com") { ... }
+
+if (evt.origin != "http://example.com") { ... }
+
+if (evt.origin !== "http://example.com") { ... }
+
+if (someRegex.test(evt.origin)) { ... }
+```
+<br />
+We have our spec, we know what to flag and more importantly, what not to. Time to start working on our rule!
+
+## Writing our rule
+
+Let's take care of the `inline` functions first since they don't need external context:
+
+#### Inline functions
+
+```yaml
+patterns:
+- pattern: |
+  window.addEventListener('message', $FUNC, ...)
+```
+
+`$FUNC` will capture the inline function into a [metavariable](https://semgrep.dev/docs/writing-rules/pattern-syntax/#metavariables). Now we can use the handy [metavariable-pattern](https://semgrep.dev/docs/writing-rules/rule-syntax/#metavariable-pattern) to exclude our **safe patterns**.
+
+```yaml
+patterns:
+- pattern: |
+    window.addEventListener('message', $FUNC, ...)
+- metavariable-pattern:
+    metavariable: $FUNC
+    patterns:
+      - pattern: |
+          function($OBJ) { ... }
+      - pattern-not: |
+          function($OBJ){ if ($OBJ.origin == $X) { ... } }
+      - pattern-not: |
+          function($OBJ){ if ($OBJ.origin === $X) { ... } }
+      - pattern-not: |
+          function($OBJ){ if ($OBJ.origin != $X) { ... } }
+      - pattern-not: |
+          function($OBJ){ if ($OBJ.origin !== $X) { ... } }
+      - pattern-not: |
+          function($OBJ){ if ($REGEX.test($OBJ.origin)) { ... } }
+```
+
+We want to match function declarations (`pattern`) that **do not** contain any of our safe patterns (`pattern-not`). A few things are worth noting here:
+
+1. Our metavariable is matching on the context captured by `$FUNC`
+2. We need the the `function($OBJ) { ... }` check to exclude external function declarations (our second scenario)
+3. The `$OBJ` metavariable is matching the function parameter and making sure the `.origin` is being called on it
+
+We could likely use a `metavariable` for our equality operators instead of declaring all of these `pattern-not`, but I like how explicit the rule currently is. Time to match those external functions!
+
+#### External functions
+
+```yaml
+pattern-either:
+  - pattern: |
+      function $FNAME(...) { $CTX }
+      ...
+      window.addEventListener('message', $FNAME,...)
+  - pattern: |
+      $FNAME = (...) => { $CTX }
+      ...
+      window.addEventListener('message', $FNAME,...)
+```
+
+_This is the only situation where I couldn't use the same pattern to match `arrow functions` and normal `function` declarations. Let me know if I could simplify this pattern._
```

Comment:
Yeah, I spent quite some time trying to figure it out and went with the `pattern-either` at the end. 😄 I wonder why in this instance normal and arrow function declarations are consider distinct. Is it because we are trying to parse a multi-line scope?